content/en/docs/features/jetstream.md

```diff
@@ -44,8 +46,24 @@ Each event is compressed by [snappy](https://github.com/google/snappy) and encod
 
 #### Resources events
 
+The resources events are published for each resource of the device. The event is published multiple times to different subjects:
+
+- once for resource ID
+- once for resource href
+- once for each resource type of the resource
+
+##### Resources events by resource ID
+
 - `plgd.owners.{ownerID}.devices.{deviceID}.resources.{resourceID}.{eventType}` publishes resources events of types `resourcechanged`, `resourcecreated`, `resourcecreatepending`, `resourcedeleted` `resourcedeletepending`, `resourceretrieved`, `resourceretrievepending`, `resourcestatesnapshottaken`, `resourceupdated`, `resourceupdatepending` for resource `resourceID`, `deviceID` and `ownerID`.
 
+##### Resources events by href
+
+- `plgd.href.owners.{ownerID}.devices.{deviceID}.resources.{resourceHrefID}.{eventType}` publishes resources events of types `resourcechanged`, `resourcecreated`, `resourcecreatepending`, `resourcedeleted` `resourcedeletepending`, `resourceretrieved`, `resourceretrievepending`, `resourcestatesnapshottaken`, `resourceupdated`, `resourceupdatepending` for resource `resourceHrefID`, `deviceID` and `ownerID`.
+
+##### Resources events by resource type
+
+- `plgd.resourceType.owners.{ownerID}.devices.{deviceID}.resourceTypes.{resourceTypeID}.{eventType}` publishes resources events of types `resourcechanged`, `resourcecreated`, `resourcecreatepending`, `resourcedeleted` `resourcedeletepending`, `resourceretrieved`, `resourceretrievepending`, `resourcestatesnapshottaken`, `resourceupdated`, `resourceupdatepending` for resource `resourceTypeID`, `deviceID` and `ownerID`.
```

Comment:
I assume this heavily affects the performance of the system. Assume you have a device, with 50 resources, each resource has 1-3 resource types. Worst case, 150 more publishes to 150 unique subjects. And this is just one device and one owner. Let's scale to 10 owners, with 10 devices -> 15k+ publishes. This won't scale well.

Reply:
So there is two ways how to achieve resource-type notification:

1. as is described. Publishing messages in RA is very cheap against update resource in DB. CPU usage is used only when command occurs.
2. gateways subscribe to all resources events and they will be filtered by resource types. In this case, all resource events need to be extended about resource types property. Otherwise gateways need to look DB. So CPU usage will be consumed by gateways for filtering out events. From point of view this is much expensive than first option.

This feature was required by infinera also by kistler so I will be fine if we found solution for  them.

Reply:
There is a big difference in 1 and 2.

Proposal n1 publishes with 10 owners and 10 devices 15k+ messages always when the device connects, when the device changes the value even though nobody is interested in. Probably it's not CPU heavy operation, but huge load on network, what will ofc increase also CPU load.

Proposal n2 - assumption is wrong that you subscribe to all resources. It will subscribe only when the user is interested in a resource type based subscription, and look-up is one-time operation.

The complexity of these two approaches is in my opinion diametrally different.

Reply:
n2) - but you need to lookup dynamically so for every publish/unpublish you will register/unregister to topics:
so if owner want's to be notified about `oic.wk.d` resource of all devices,  - gateways needs to:

1. subscribe to `plgd.owners.{ownerID}.devices.*.resource-links.>`
2. lookup to DB for resource-links with resource type filter (if there is 10k devices this will be slow because it will be fulltext search)
3. subscribe/unsubscribe per resource
4. if event occurs from `plgd.owners.{ownerID}.devices.*.resource-links.>` use step 3

Reply:
Indexes over array are not working in mongodb ...

Reply:
This could be expensive in memory in gateway: they are registered per resource - so the consumption of memory could be big for per subscription to NATs.

Reply:
It needs to be verified. But this is the easiest verification - how expensive is the subscription through an already opened NATS connection.

Reply:
We cannot use this approach. Between step 4 and 3 you can loose events. For example new resource was added but ResourceChanged was lost because the it takes some time ...

So you need to use my `n2` without dynamic registration.

Reply:
You can reconcile. Event was lost but you subscribe and pull from RD. Same as you do for pending commands. Database and it’s cache is there to read data from it. This scenario is not typical, happening in regular operation of live devices. It’s only when new devices are connected or existing reconnected. 

Look, Let’s do 2 tests. Subscription one, I wrote above, and then, update rd to publish 512 more events to random subjects per device. And run infn use case with 10k devices. 

Then we can discuss more concrete points and what and where to optimize.

Reply:
OK. I found the solution. If gateway need to be notified on event it will create record in DB about subject.  RA will read records and streams to that subjects when event was created. Something similar to router.

Reply:
You are loosing authorization context if it is driven by the gateway. And as you wrote, you are doing router from CGW, making him aware of infrastructure, which you model in form of subjects. That’s not responsibility of RA. 

Let’s do performance tests of 2 proposals and then elaborate further. There is no pressure. We have time, we need tangible data and think about it.

Reply:
The router is responsible for publish events to nats. So currently there is a basic router that stream just one subject.
Ofcourse you can still organize subjects in with auth context. I m against point 2.